include/astra_camera/astra_driver.h

```diff
@@ -69,11 +69,79 @@
 
 #include <ros/ros.h>
 
+#include <actionlib/server/simple_action_server.h>
+#include <camera_control_msgs/GrabImagesAction.h>
+
+#include <diagnostic_updater/diagnostic_updater.h>
+#include <diagnostic_updater/publisher.h>
+
 namespace astra_wrapper
 {
 
+struct CameraParameters
+{
+  int exposure;
+  int gain;
+  bool auto_exposure;
+};
+
+/**
+ * @brief The ActionImageSyncer struct
+ * @note When the change in a parameter of the camera is big, the camera is
+ * slow to react and won't give the desired image. Skipping frames accommodates
+ * for this. Also, keep in mind that the camera is queried about a parameter, it
+ * responds with the set value and not the currently reached one.
+ */
+struct ActionImageSyncer
+{
+  ActionImageSyncer() :
+      num_skip_frames_default(2),
+      frames_skipped(0),
+      need_image(false),
+      acquired(true)
+  {}
+
+  void reset(bool skip_frames)
+  {
+    num_skip_frames = skip_frames ? num_skip_frames_default : 0;
+    frames_skipped = 0;
+    need_image = true;
+    acquired = false;
+  }
+
+  void check()
+  {
+    if (frames_skipped == num_skip_frames)
+    {
+      need_image = false;
+      acquired = true;
```

Comment:
why do we acquire here anything? its seems not to be related if we have the image or not

Reply:
Getting one image is not considered as an acquired state. We need to skip some frames before we go into that state. The auto-brightness of the device needs more than one image.

Reply:
that the issue with this code (and one of the causes of many issues below)
why don't you define a 
```
void update(const sensos_msgs::ImagePtr& _img) {
    if(need_image && frames_skipped++ == num_skip_frames){
        boost::lock_guard lk(mutex);
        image = _img;
        acquired = true;
      cv.notify_one();
    }
}
```
and use it in your image callback

Reply:
The subscriber logic interferes here. needs_image is accessed again in the image callback. So the mutex is locked once in the callback, and then the syncer is accessed.